kusto/ingest/ingest.go

```diff
@@ -440,7 +441,87 @@ var (
 	ErrTooLarge = errors.ES(errors.OpIngestStream, errors.KClientArgs, "cannot add data larger than 4MiB")
 )
 
-const mib = 1024 * 1024
+const (
+	MaxRetryCount    = 3
+	mib              = 1024 * 1024
+	MaxStreamingSize = 4 * mib
+)
+
+// FromFile allows uploading a data file for Kusto from either a local path or a blobstore URI path.
+// This method is thread-safe.
+func (i *Ingestion) FromFileManaged(ctx context.Context, fPath string,
+	options ...FileOption) (*Result, error) {
+	result, props, err := i.prepForIngestion(ctx, options)
+	if err != nil {
+		return nil, err
+	}
+
+	result.record.IngestionSourcePath = fPath
+
+	local, err := filesystem.IsLocalPath(fPath)
+	if err != nil {
+		return nil, err
+	}
+
+	skipStreaming := false
+
+	if local {
+		stat, err := os.Stat(fPath)
+		if err != nil {
+			return nil, err
+		}
+		skipStreaming = stat.Size() > MaxStreamingSize
+	}
+
+	if skipStreaming { // In case of blob it's inefficient to use streaming ingest,
+		// so we're defaulting to queued
+		if local {
+			err = i.fs.Local(ctx, fPath, props)
+		} else {
+			err = i.fs.Blob(ctx, fPath, 0, props)
+		}
+		if err != nil {
+			return nil, err
+		}
+		result.putQueued(i.mgr)
+		return result, nil
+	}
+
+	file, err := os.Open(fPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return i.FromReaderManaged(ctx, file, options...)
+}
+
+func (i *Ingestion) FromReaderManaged(ctx context.Context, reader io.Reader,
+	options ...FileOption) (*Result, error) {
+	result, props, err := i.prepForIngestion(ctx, options)
+	if err != nil {
+		return nil, err
+	}
+
+	bytesRead, err := ioutil.ReadAll(reader)
```

Comment:
The unknown size of blobs escalates here

Reply:
The problem is that the interface doesn't let us to get the size before hand.
I'll look into properties or other methods

Reply:
Now I use the props, and peek to never actually read more than the max size. What do you think?